synapse/api/errors.py

```diff
@@ -86,6 +87,11 @@ class CodeMessageException(RuntimeError):
 
     def __init__(self, code, msg):
         super(CodeMessageException, self).__init__("%d: %s" % (code, msg))
+
+        # Convert http.HTTPStatus objects to their int status code equivalents
```

Comment:
ohh gosh, this is gnarly.

so it turns out that HTTPStatus is a subclass of int:

```
>>> import http
>>> isinstance(http.HTTPStatus.FORBIDDEN, int)
True
```

... so the docstring is correct. Indeed everything works correctly except for the access logger, where it does `str(code)`. Twisted does `b"%d" % code` which does the right thing too.

Incidentally, the change of `http.client.<code>` from `int` to `IntEnum` happened in python 3.5.

It feels like there are multiple potential solutions here:

* go around replacing every `HTTPStatus` reference with `<code>.value` (which may be tricky, because there are many different ways of getting hold of an HTTPStatus reference - which is a separate problem; we should be consistent.)
* do this and special-case `HTTPStatus` here
* do `code = int(code)` here
* do `code = "%d" % (self.code,)` in the access logger code

I'm not sure I have strong feelings. I'm slightly uneasy about the last option, which feels like it might be a bit fragile if Twisted ever decides to switch from `b"%d" % code` to `str(code).encode()` or something.

Either way I'd suggest expanding the comments to explain that `HTTPStatus` instances are `int`s but with magic `__str__` methods which mess us up.

Reply:
Thank you for the detailed comment! It looks like we're leaning towards option 3, bearing in mind to aim for option 1 in the long term.

I'll go ahead and change the PR to reflect that and update the comments.